finality-aleph/src/validator_network/mod.rs

```diff
@@ -0,0 +1,69 @@
+#![allow(dead_code)]
+use std::fmt::Display;
+
+use aleph_primitives::AuthorityId;
+use codec::Codec;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+mod io;
+#[cfg(test)]
+mod mock;
+
+/// What the data sent using the network has to satisfy.
+pub trait Data: Clone + Codec + Send + Sync + 'static {}
+
+impl<D: Clone + Codec + Send + Sync + 'static> Data for D {}
+
+/// Network represents an interface for opening and closing connections with other Validators,
+/// and sending direct messages between them.
+///
+/// Note on Network reliability and security: it is neither assumed that the sent messages must be
+/// always delivered, nor the established connections must be secure in any way. The Network
+/// implementation might fail to deliver any specific message, so messages have to be resent while
+/// they still should be delivered.
+#[async_trait::async_trait]
+pub trait Network<A: Data, D: Data>: Send {
```

Comment:
Can we move those `D` and `A` parameters on per method basis. A single instance of network is able to send anything fullfiling `D`. It can simplify things like protocol versioning - you can just start sending new type of messages, without the need to wrap your types, or some other hackery.

Reply:
Unfortunately that is not the case, this interface will be implemented as a set of channels that pass the messages and other calls to a running service. These channels have to be parametrized by `D`. :/